### DISCUSSION TITLE
Help with ColumnView

### DISCUSSION THREAD

**Discussion by RegProctor** [diamondburned/gotk4-examples#35] [Help] [answered]
I started writing to implement a ColumnView and as I got further along it looked like it needed a couple interface implementations (maybe more if I kept going down the rabbit hole)  so I thought maybe I missed something that would make this considerably easier, like some built-in functions that implement these as standard for common objects.

Could someone have a look and let me know if this seems right or if I went awry somewhere?

I stopped at line 72:  `lm := gio.ListModeller()` as that seemed like where it was starting to go down the rabbit hole. I don't mind implementing the interfaces down the line if that's what's needed but I suspect there is a simpler way.

This is just proof of concept code so it's a pretty simple. I borrowed from https://ssalewski.de/gtkprogramming.html#_listview_columnview_gridview_and_dropdown for ideas on how to approach the code.

```go
package gui

import (
	`strconv`
	`time`

	`github.com/diamondburned/gotk4/pkg/gio/v2`
	`github.com/diamondburned/gotk4/pkg/glib/v2`
	`github.com/diamondburned/gotk4/pkg/gtk/v4`

	`gitlab.com/<my DB Layer>/dbLayer`
)

type Scroll struct {
	*gtk.ScrolledWindow
}

type GtkColumnViewDaysAndTimes struct {
	app *App
	id  string
	cv  *gtk.ColumnView
	*dbLayer.GUI
	saveTimer    *time.Timer
	settingValue bool
}

func NewGtkColumnViewDaysAndTimes(id string, a *App) (dat *GtkColumnViewDaysAndTimes) {
	cv := a.builder.GetObject(id).Cast().(*gtk.ColumnView)
	dat = &GtkColumnViewDaysAndTimes{
		a,
		id,
		cv,
		a.db,
		nil,
		false,
	}
	dat.setup()
	return
}

func (dat *GtkColumnViewDaysAndTimes) setup() {
	f0 := gtk.NewSignalListItemFactory()
	f0.ConnectSetup(dat.guiSignalSetupC0)
	f0.ConnectBind(dat.guiSignalBindC0)
	c0 := gtk.NewColumnViewColumn(`Apply`, &f0.ListItemFactory)
	dat.cv.AppendColumn(c0)

	f1 := gtk.NewSignalListItemFactory()
	f1.ConnectSetup(dat.guiSignalSetupC1)
	f1.ConnectBind(dat.guiSignalBindC1)
	c1 := gtk.NewColumnViewColumn(`Timezone`, &f1.ListItemFactory)
	dat.cv.AppendColumn(c1)

	types := []glib.Type{
		glib.TypeBoolean,
		glib.TypeString,
	}

	ls := gtk.NewListStore(types)

	for i := 1; i < 5; i++ {
		ls.Set(
			ls.Append(),
			[]int{0, 1},
			[]glib.Value{
				*glib.NewValue(i%2 == 0),
				*glib.NewValue(`A comment ` + strconv.Itoa(i)),
			},
		)
	}

	lm := gio.ListModeller()
	sm := gtk.NewSingleSelection(lm)
	dat.cv.SetModel(sm)

}

func (dat *GtkColumnViewDaysAndTimes) guiSignalSetupC0(listItem *gtk.ListItem) {
	child := gtk.NewCheckButton()
	listItem.SetChild(child)
}

func (dat *GtkColumnViewDaysAndTimes) guiSignalSetupC1(listItem *gtk.ListItem) {
	child := gtk.NewEntry()
	listItem.SetChild(child)
}

var count = 1

func (dat *GtkColumnViewDaysAndTimes) guiSignalBindC0(listItem *gtk.ListItem) {
	child := listItem.Child().(*gtk.CheckButton)
	child.SetLabel(`Count ` + strconv.Itoa(count))
	count++
}

func (dat *GtkColumnViewDaysAndTimes) guiSignalBindC1(listItem *gtk.ListItem) {
	child := listItem.Child().(*gtk.Entry)
	child.SetText(`Also ` + strconv.Itoa(count))
}
```

**Comment by diamondburned** [comment]
> I don't mind implementing all the interfaces down the line if that's what's needed but I suspect there is a simpler way.

You unfortunately cannot even do this. Doing so would require subclassing support which gotk4 currently lacks.

I recommend using `gtk.StringList` and maintain a `map[string]T`. See [this particular file in the commit](https://github.com/diamondburned/gtkcord4/commit/2f8a342074f420ef5a21dfb74e8c389d5d77d8ba#diff-bbfe43ebad09fe6f39521b6dff996cbe9bacaac830644a9a406440183b417372R158) for an example in gtkcord4.

**Reply by RegProctor** [reply]
Thanks for the quick reply, I'll go down that path instead.

**Reply by RegProctor** [reply]
I spoke a little too soon. That really is just list of strings. I need a row with a checkbox and 2 or 3 entry boxes and the rows need to be dynamic since I'll be pulling them from a DB and they will grow and shrink over time. Perhaps GTK and Go together isn't good fit for this yet. I'll do what digging I can and see if I can make this work but if you have any thoughts please do tell. I'm wondering if I can bypass some of the problems if I initially suck up the structure from a UI file instead of programmatically creating it.

Any thoughts you have appreciated as I am brand new to GTK (about a week into it), it's also my first and only GUI experience with Go, and in go I'm okay but certainly not seasoned with it yet.

**Reply by RegProctor** [reply]
I'm wondering if the reason you are finding the documentation on subclasses inadequate is because they are moving away from it in favor of delegation—I came across this blog post [Custom widgets in GTK 4 – Introduction](https://blog.gtk.org/2020/04/23/custom-widgets-in-gtk-4-introduction/) and perhaps a different way of thinking about the problem would help?

**Reply by RegProctor** [reply]
One last question. You see that I create a listStore with `ls := gtk.NewListStore(types)` but I can't see how to get that into the ColumnView. Is that part of you needing subclassing or is that just me not knowing enough?  If the later could you show me how and I'll play around with it to see what I can do from there.

In the meantime I'm going to assume there's no good way to use a ColumnView for now and go down your route and just build what I need from the simpler widgets.

**Reply by diamondburned** [reply]
> That really is just list of strings. I need a row with a checkbox and 2 or 3 entry boxes and the rows need to be dynamic since I'll be pulling them from a DB and they will grow and shrink over time.

The widgets you need aren't relevant to the list store itself. That's the job of the widget factory (`SignalListItemFactory`). Ideally, the list store should actually just contain just enough information to identify *rows* apart. Those identifiers can then be used differently depending on the factory.

**Reply by diamondburned** [reply]
> but I can't see how to get that into the ColumnView

You would want to create a new `SelectionModel` and give its constructor your `ListModel`. Then you can give `ColumnView` the new `SelectionModel`.

**Reply by jgillich** [reply]
Having the same problem as OP; is it even possible to use StringList with ColumnView?

```go
list := gtk.NewStringList([]string{"foo", "bar", "baz", "inga"})
cv := gtk.NewColumnView(gtk.NewSingleSelection(list))
cv.AppendColumn(gtk.NewColumnViewColumn("Name", &gtk.NewSignalListItemFactory().ListItemFactory))
cv.AppendColumn(gtk.NewColumnViewColumn("Namespace", &gtk.NewSignalListItemFactory().ListItemFactory))
```

Renders 4 blank columns

![Screenshot from 2024-01-17 09-42-03](https://github.com/diamondburned/gotk4-examples/assets/347965/bf3a04ba-5e3b-40c1-8b8c-3d4fe3ef732e)

And I'm just as lost on how to get this working with `ListStore`. Are there any examples for creating a `SelectionModel`? :pray:

**Reply by jgillich** [reply]
Ah, looks like you need to render it via the factory. Probably not ideal but I got it working with this:

```go
columns := []string{"Name", "Namespace"}

for i, name := range columns {
	ii := i
	factory := gtk.NewSignalListItemFactory()
	factory.ConnectBind(func(listitem *gtk.ListItem) {
		s := listitem.Item().Cast().(*gtk.StringObject).String()
		label := gtk.NewLabel(strings.Split(s, "|")[ii])
		label.SetHAlign(gtk.AlignBaseline)
		listitem.SetChild(label)
	})
	column := gtk.NewColumnViewColumn(name, &factory.ListItemFactory)
	cv.AppendColumn(column)
}
```

With string list, you only get one string for the whole row, that's why I split it for each column.

If found a [rust example](https://davecompton.net/2022/05/01/rust-gtk4-columnview.html) that binds the properties in the setup handler, not sure if we can do something similar in gotk.

**Reply by diamondburned** [reply]
I'm not aware of anything like binding, but you can do so by doing

```go
sobj.NotifyProperty("string", func() { label.SetText(sobj.String()) })
```

This is of course only necessary if the `StringObject` is mutable. A slightly better way of abusing the `StringObject` would be to encode it as JSON.

I'm also experimenting with using `core/gbox` to encode a fake pointer in there to get a Go object, but this of course means that changes in the Go model won't be reflected in the actual model itself.

**Comment by RegProctor** [comment, marked as answer]
> create a new `SelectionModel` and give its constructor your `ListModel`. Then you can give `ColumnView` the new `SelectionModel`

I figure that even though I was looking at other possibilities. The problem has been that getting a definition of `ListModel` for `ColumnView` that made sense was proving to be problematic, more so than it seems it should be but then I did a search for `*gio.ListModel` to see all of what was returning it and one of the results was: 

      func (self *ColumnView) Columns() *gio.ListModel {

I think that was the key I was looking for as I can plug that straight into a selection model creation and then that selection model goes back into the `ColumnView` and since the `gio.ListModel` came from the `ColumnView` in the first place, it now seems like the `SelectionModel` must be created with all the correct column information. It's late so I'll have to follow this up with testing in the evening but on the surface that looks like that is the key to the step I was concerned about.

BTW, I agree with a comment you made in your discord repo. at some point, _"ColumnView is hard",_ or at the very least could be a lot easier but GTK was built in C and C & Go don't necessary mesh well when it come to methodologies crossing over so I guess we just got to _"suck it up"_ at times in this case to use an old expression.

**UPDATE:** That was wrong as I didn't realize they were also dealing with columns in terms of list models. You can't feed a column model into something for rows/data. Now that I understand it I saw where the subclassing is needed. I'm just going to build the ColumnView on the go side until there's a way to do this. Thanks.

For anyone who gets to this, the subclassing is needed because there is one list for all the columns of data, not a list per column but there is only finite types you can pass to gio.NewListStore(), not to mention only one parameter. This means you have to create a struct with all the data types and then you have register that struct as a new type with GTK. Once it's registered with GTK then you can pass it to gio.NewListStore() and without subclassing this isn't going to happen.
